Jellyfin.Server/Migrations/Routines/MigrateKeyframeData.cs

```diff
@@ -0,0 +1,172 @@
+    private readonly ILogger<MoveTrickplayFiles> _logger;
+    private readonly IApplicationPaths _appPaths;
+    private readonly IDbContextFactory<JellyfinDbContext> _dbProvider;
+    private static readonly JsonSerializerOptions _jsonOptions = JsonDefaults.Options;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="MigrateKeyframeData"/> class.
+    /// </summary>
+    /// <param name="libraryManager">Instance of the <see cref="ILibraryManager"/> interface.</param>
+    /// <param name="logger">The logger.</param>
+    /// <param name="appPaths">Instance of the <see cref="IApplicationPaths"/> interface.</param>
+    /// <param name="dbProvider">The EFCore db factory.</param>
+    public MigrateKeyframeData(
+        ILibraryManager libraryManager,
+        ILogger<MoveTrickplayFiles> logger,
+        IApplicationPaths appPaths,
+        IDbContextFactory<JellyfinDbContext> dbProvider)
+    {
+        _libraryManager = libraryManager;
+        _logger = logger;
+        _appPaths = appPaths;
+        _dbProvider = dbProvider;
+    }
+
+    private string KeyframeCachePath => Path.Combine(_appPaths.DataPath, "keyframes");
+
+    /// <inheritdoc />
+    public Guid Id => new("EA4bCAE1-09A4-428E-9B90-4B4FD2EA1B24");
+
+    /// <inheritdoc />
+    public string Name => "MigrateKeyframeData";
+
+    /// <inheritdoc />
+    public bool PerformOnNewInstall => false;
+
+    /// <inheritdoc />
+    public void Perform()
+    {
+        const int Limit = 100;
+        int itemCount = 0, offset = 0, previousCount;
+
+        var sw = Stopwatch.StartNew();
+        var itemsQuery = new InternalItemsQuery
+        {
+            MediaTypes = [MediaType.Video],
+            SourceTypes = [SourceType.Library],
+            IsVirtualItem = false,
+            IsFolder = false
+        };
+
+        using var context = _dbProvider.CreateDbContext();
+        using var transaction = context.Database.BeginTransaction();
+        List<KeyframeData> keyframes = [];
+
+        do
+        {
+            var result = _libraryManager.GetItemsResult(itemsQuery);
+            _logger.LogInformation("Importing keyframes for {Count} items", result.TotalRecordCount);
+
+            var items = result.Items;
+            previousCount = items.Count;
+            offset += Limit;
+            foreach (var item in items)
+            {
+                if (TryGetKeyframeData(item, out var data))
+                {
+                    keyframes.Add(data);
+                }
+
+                if (++itemCount % 10_000 == 0)
+                {
+                    context.KeyframeData.AddRange(keyframes);
+                    keyframes.Clear();
+                    _logger.LogInformation("Imported keyframes for {Count} items in {Time}", itemCount, sw.Elapsed);
+                }
+            }
+        } while (previousCount == Limit);
+
+        context.KeyframeData.AddRange(keyframes);
+        context.SaveChanges();
+        transaction.Commit();
```

Comment:
It's true that in BaseItemRepository's DeleteItem everything is wrapped in a transaction as well, but afaik this is unnecessary: SaveChanges does already everything in a transaction, transaction only make sense if you want to call SaveChanges multiple times and make all of them succeed or fail together.

Correct me if I'm wrong.

Reply:
You're correct, the initial implementation called save multiple times, so this was necessary. It doesn't hurt to be there though